pkg/controller/clustercatalog_controller.go

```diff
@@ -0,0 +1,401 @@
+}
+
+// RequirementsEnsurer is an abstraction that is used to represent a
+// type that has a method to ensure that required fields are set as expected on a provided
+// *unstructured.Unstructured
+type RequirementsEnsurer interface {
+	// EnsureRequirements ensures that pre-configured requirements are present on the
+	// provided *unstructured.Unstructured object, setting them if they are not present
+	// or the value does not match the requirement. It returns:
+	//   - A boolean that represents whether or not the provided *unstructured.Unstructured was modified (true == modified)
+	//   - An error if one occurred during the
+	EnsureRequirements(*unstructured.Unstructured) (bool, error)
+
+	// Required returns an *unstructured.Unstructured object matching all the requirements
+	Required() *unstructured.Unstructured
+}
+
+type metadataSpecEnsurer struct {
+	required *unstructured.Unstructured
+}
+
+func (mse *metadataSpecEnsurer) Required() *unstructured.Unstructured {
+	return mse.required.DeepCopy()
+}
+
+func (mse *metadataSpecEnsurer) EnsureRequirements(existing *unstructured.Unstructured) (bool, error) {
+	requiredCopy := mse.required.DeepCopy()
+	metadataModified, err := ensureRequiredMetadata(requiredCopy, existing)
+	if err != nil {
+		return false, fmt.Errorf("comparing required and existing metadata for ClusterCatalog %q: %w", mse.required.GetName(), err)
+	}
+
+	specModified, err := ensureRequiredSpec(requiredCopy, existing)
+	if err != nil {
+		return metadataModified, fmt.Errorf("comparing required and existing spec for ClusterCatalog %q: %w", mse.required.GetName(), err)
+	}
+
+	return metadataModified || specModified, nil
+}
+
+type Applier interface {
+	Apply(context.Context, *unstructured.Unstructured, *unstructured.Unstructured) error
+}
+
+type applier struct {
+	client   dynamic.Interface
+	gv       schema.GroupVersion
+	kind     string
+	resource string
+	name     string
+}
+
+func (a *applier) Apply(ctx context.Context, original, modified *unstructured.Unstructured) error {
+	// do nothing if modified is not provided
+	if modified == nil {
+		return nil
+	}
+	// Do a simple create if there is no original
+	if original == nil {
+		_, err := a.client.Resource(a.gv.WithResource(a.resource)).Create(ctx, modified, metav1.CreateOptions{})
+		if err != nil && !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("creating resource %s %q: %w", a.gv.WithResource(a.resource), modified.GetName(), err)
+		}
+		return nil
+	}
+
+	// Generate and apply a patch for the diff between original and modified
+	diffPatch, err := generatePatch(original, modified)
+	if err != nil {
+		return fmt.Errorf("generating patch: %w", err)
+	}
+
+	_, err = a.client.Resource(a.gv.WithResource(a.resource)).Patch(ctx, original.GetName(), types.MergePatchType, diffPatch, metav1.PatchOptions{
```

Comment:
Note: After a discussion with @joelanford I'm going to spend some more time looking to see if we can simply use Server Side Apply instead of attempting to do a merge patch.

I originally ran into a swath of issues trying to do SSA but Joe shared a couple more resources with me that may be helpful in resolving some of the issues I encountered.

Reply:
+1 to server side apply.